File: app/src/main/java/com/github/alextokarew/javadebugging/ffm/FibFFM.java
package com.github.alextokarew.javadebugging.ffm;

import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;

public class FibFFM {
    static void main(String[] args) throws Throwable {
        try (Arena arena = Arena.ofConfined()) {
            IO.println("Working dir: " + System.getProperty("user.dir"));
            var sl = SymbolLookup.libraryLookup("../cpp-fib/build/lib/main/debug/libcpp-fib.so", arena);
            MemorySegment memSeg = sl.find("fibonacci").orElseThrow();
            Linker linker = Linker.nativeLinker();

            var desc = FunctionDescriptor.of(ValueLayout.JAVA_INT, ValueLayout.JAVA_INT);

            MethodHandle fibonacci = linker.downcallHandle(memSeg, desc);
            var result = fibonacci.invoke(12);
            IO.println("Result: " + result);
        }
    }
}

File: app/src/main/cpp/fibffm.cpp
#include <iostream>
#include "fibffm.h"

int fibonacci(int n) {
    std::cout << "Start ffm fibonacci" << std::endl;
    int prev = 0, result = 1;
    for (int i = 1; i < n; i++) {
        int old_result = result;
        result = result + prev;
        prev = old_result;
    }
    std::cout << "Finish ffm fibonacci" << std::endl;
    return result;
}
